dali/kernels/signal/fft/fft_postprocess.cuh

```diff
@@ -0,0 +1,361 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SIGNAL_FFT_FFT_POSTPROCESS_CUH_
+#define DALI_KERNELS_SIGNAL_FFT_FFT_POSTPROCESS_CUH_
+
+#include <cuda_runtime.h>
+#include "dali/core/geom/vec.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/signal/fft/fft_common.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+namespace fft_postprocess {
+
+using namespace fft;  // NOLINT
+
+template <typename Out, typename In>
+struct SampleDesc {
+  Out *out;
+  const In *in;
+  ptrdiff_t out_stride;
+  ptrdiff_t in_stride;
+};
+
+struct BlockDesc {
+  int sample_idx;
+  ivec2 start, end;
+};
+
+constexpr int kBlock = 32;
+
+struct norm2square {
+  DALI_HOST_DEV DALI_FORCEINLINE
+  auto operator()(float2 c) const {
+    return c.x * c.x + c.y * c.y;
+  }
+};
+
+struct norm2 {
+  DALI_HOST_DEV DALI_FORCEINLINE
+  auto operator()(float2 c) const {
+#ifdef __CUDA_ARCH__
+    return sqrtf(c.x * c.x + c.y * c.y);
+#else
+    return std::sqrt(c.x * c.x + c.y * c.y);
+#endif
+  }
+};
+
+struct power_dB {
+  power_dB() = default;
+  power_dB(float cutoff_dB) {
+    cutoff = std::pow(10, cutoff_dB / 10);
+  }
+
+  float mul = 3.01029995664f;  // log10(2)
+  float cutoff = 1e-8;         // -80 dB
+
+  DALI_HOST_DEV DALI_FORCEINLINE
+  auto operator()(float2 c) const {
+#ifdef __CUDA_ARCH__
+    return mul * log2f(::max(c.x * c.x + c.y * c.y, cutoff));
+#else
+    return mul * std::log2(std::max(c.x * c.x + c.y * c.y, cutoff));
+#endif
+  }
+};
+
+template <typename Out, typename In, typename Convert>
+__global__ void ConvertTimeMajorSpectrogram(
+      Out *out, int out_stride,
+      const In *in, int in_stride, int nfft, int64_t nwindows,
+      Convert convert = {}) {
+  // A warp processes a whole row (transform) to ensure sequential processing
+  // and thus enable in-place operation.
+
+  int64_t wnd = static_cast<int64_t>(blockIdx.y) * blockDim.y + threadIdx.y;
+  if (wnd >= nwindows)
+    return;
+  out += wnd * out_stride;
+  in += wnd * in_stride;
+
+  // The loop starts at 0 (not threadIdx.x) to ensure there's no divergence which would
+  // cause undefined behavior in __syncwarp() (or require complex mask calculation).
+  for (int i = 0; i < nfft; i += 32) {
+    int j = i + threadIdx.x;
+    Out v = j < nfft ? convert(in[j]) : Out();
+    __syncwarp();  // memory barrier for in-place execution
```

Comment:
You can add comment that this applies for the case sizeof(Out) != sizeof(In) and inplace calculation.

Reply:
And you can skip `__syncwarp` if this is not an inplace case.